piptools/writer.py

```diff
@@ -230,6 +232,8 @@ def _format_requirement(
         ireq_hashes = (hashes if hashes is not None else {}).get(ireq)
 
         line = format_requirement(ireq, marker=marker, hashes=ireq_hashes)
+        if self.constraints_compatible:
+            line = re.sub(r"\[.*\]", "", line)
```

Comment:
Maybe something like this?
```suggestion
            line = re.sub(r"\[[^\[\]+]\]", "", line)
```

Reply:
use non-greedy regex: 
`r"\[.+?\]"`